### ISSUE TITLE
Layer Normalization

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/deep-learning-readinglist#154]
https://arxiv.org/pdf/1607.06450.pdf

**Comment by sweaterr** [comment]
https://twimlai.com/new-layer-normalization-technique-speeds-rnn-training/

**Comment by sweaterr** [comment]
http://davidstutz.de/layer-normalization-ba-et-al/

**Comment by sweaterr** [comment]
https://tensorflow.blog/2016/07/24/layer-normalization/

**Comment by sweaterr** [comment]
## abstract

Training state-of-the-art, deep neural networks is computationally expensive. One way to reduce the training time is to normalize the activities of the neurons. A recently introduced technique called batch normalization uses the distribution of the summed input to a neuron over a mini-batch of training cases to compute a mean and variance which are then used to normalize the summed input to that neuron on each training case. This significantly reduces the training time in feedforward neural networks. However, the effect of batch normalization is dependent on the mini-batch size and it is not obvious how to apply it to recurrent neural networks. In this paper, we transpose batch normalization into layer normalization by computing the mean and variance used for normalization from all of the summed inputs to the neurons in a layer on a single training case. Like batch normalization, we also give each neuron its own adaptive bias and gain which are applied after the normalization but before the non-linearity. Unlike batch normalization, layer normalization performs exactly the same computation at training and test times. It is also straightforward to apply to recurrent neural networks by computing the normalization statistics separately at each time step. Layer normalization is very effective at stabilizing the hidden state dynamics in recurrent networks. Empirically, we show that layer normalization can substantially reduce the training time compared with previously published techniques.

**Comment by sweaterr** [comment]
## 1 Introduction 

Deep neural networks trained with some version of Stochastic Gradient Descent have been shown to substantially outperform previous approaches on various supervised learning tasks in computer vision [Krizhevsky et al., 2012] and speech processing [Hinton et al., 2012]. 

But state-of-the-art deep neural networks often require many days of training. 

It is possible to speed-up the learning by computing gradients for different subsets of the training cases on different machines or splitting the neural network itself over many machines [Dean et al., 2012], but this can require a lot of communication and complex software. 

It also tends to lead to rapidly diminishing returns as the degree of parallelization increases. 

An orthogonal approach is to modify the computations performed in the forward pass of the neural net to make learning easier. 

Recently, batch normalization [Ioffe and Szegedy, 2015] has been proposed to reduce training time by including additional normalization stages in deep neural networks. 

The normalization standardizes each summed input using its mean and its standard deviation across the training data. 

Feedforward neural networks trained using batch normalization converge faster even with simple SGD. In addition to training time improvement, the stochasticity from the batch statistics serves as a regularizer during training.

**Comment by sweaterr** [comment]
Despite its simplicity, batch normalization requires running averages of the summed input statistics. 

In feed-forward networks with fixed depth, it is straightforward to store the statistics separately for each hidden layer. 

However, the summed inputs to the recurrent neurons in a recurrent neural network (RNN) often vary with the length of the sequence so applying batch normalization to RNNs appears to require different statistics for different time-steps. Furthermore, batch normalizaarXiv:1607.06450v1 [stat.ML] 21 Jul 2016 tion cannot be applied to online learning tasks or to extremely large distributed models where the minibatches have to be small. This paper introduces layer normalization, a simple normalization method to improve the training speed for various neural network models. Unlike batch normalization, the proposed method directly estimates the normalization statistics from the summed inputs to the neurons within a hidden layer so the normalization does not introduce any new dependencies between training cases. We show that layer normalization works well for RNNs and imp

**Comment by sweaterr** [comment]
2 Background

![image](https://user-images.githubusercontent.com/1518919/28560912-5fc8a7ce-7157-11e7-8efc-193b39addf36.png)

**Comment by sweaterr** [comment]
We now consider the layer normalization method which is designed to overcome the drawbacks of
batch normalization. 

Notice that changes in the output of one layer will tend to cause highly correlated changes in the summed inputs to the next layer, especially with ReLU units whose outputs can change by a lot.
한 레이어의 출력이 변경되면 결과가 많이 변경 될 수있는 ReLU 유닛의 경우 특히 다음 레이어에 대한 합계 입력의 상호 관련성이 높은 변경이 발생하는 경향이 있습니다.

This suggests the “covariate shift” problem can be reduced by fixing the mean and the variance of the summed inputs within each layer. 

We, thus, compute the layer normalization statistics over all the hidden units in the same layer as follows:
